Refactor Iceberg connector to support Iceberg native catalogs

jackye1995 opened the pull request:
Had some offline discussions with the maintainers of the Iceberg connector @zhenxiao @ChunxuTang @beinan , the direction that the community would like to go is to gradually remove its Hive dependency and make a pure Iceberg connector that can evolve independently.

Based on this ask, this PR introduces a `nativeMode` that user can turn on in order to use a different `ConnectorMetadata` implementation called `IcebergNativeMetadata`. This implementation has no dependency on the Hive connector. The plan is to first develop all the features using this switch, and once it is feature complete and stable, we can drop all the `presto-hive` related dependencies and deprecate the old implementations.

Test plan: added distributed and smoke tests running using Iceberg's `HadoopCatalog`, all tests pass.

```
== RELEASE NOTES ==
Iceberg Changes
* Iceberg connector now supports a native mode that can be used without a Hive installation to run queries against Iceberg  native catalogs.
```

cc @pettyjamesm

jackye1995 commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergConnector.java:
```diff
@@ -188,10 +188,7 @@ public void commit(ConnectorTransactionHandle transaction)
     {
-        IcebergMetadata metadata = transactionManager.remove(transaction);
-        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(getClass().getClassLoader())) {
-            metadata.rollback();
```
The rollback code does nothing anyway, so I just removed it.

jackye1995 commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergConfig.java:
```diff
@@ -26,6 +29,10 @@
     private IcebergFileFormat fileFormat = PARQUET;
     private HiveCompressionCodec compressionCodec = GZIP;
+    private boolean nativeMode;
+    private String catalogImpl = HiveCatalog.class.getName();
```
Iceberg in general uses the pattern of dynamic plugin, users can plugin their own implementations of `Catalog`, `FileIO` and a couple of other important classes. If this is an anti-pattern for Presto, we can move to a Enum type. This feature is mostly oriented for customization of enterprise users, they also need to load in their own jar containing the implementation class during installation time.

jackye1995 commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergResourceFactory.java:
```diff
@@ -0,0 +1,113 @@
+        throw new PrestoException(NOT_SUPPORTED, "Iceberg catalog implementation " + catalogImpl + " does not support namespace operations");
+    }
+
+    private String getCatalogCacheKey(ConnectorSession session)
```
This is basically to get a string representation of the user identity. As long as the identity does not change, the same catalog implementation can be used. This technically should not need any cache expiration, because the Catalog instance itself does not check access, but when any catalog action is called the access of the identity is checked, so the cache can still work when the access policy of the identity changes at server side.

jackye1995 commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergResourceFactory.java:
```diff
@@ -0,0 +1,113 @@
+import static org.apache.iceberg.CatalogProperties.WAREHOUSE_LOCATION;
+
+/**
+ * Factory for loading Iceberg resources such as Catalog and FileIO.
```
FileIO related caching is not added yet, just to keep the initial implementation simple.

jackye1995 commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergSplitManager.java:
```diff
@@ -58,8 +68,14 @@ public ConnectorSplitSource getSplits(
-        Table icebergTable = getIcebergTable(metastore, hdfsEnvironment, session, table.getSchemaTableName());
+        Table icebergTable;
+        if (nativeCatalogMode) {
+            icebergTable = resourceFactory.getCatalog(session).loadTable(toIcebergTableIdentifier(table.getSchemaTableName()));
```
although this is reloading the Iceberg table, because we always input the actual snapshot ID, snapshot isolation is still guaranteed. Some optimizations can be made to cache the table object, but I am not adding that here to keep things simple.

beinan commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergNativeMetadata.java:
```diff
@@ -0,0 +1,523 @@
+    public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTableHandle tableHandle, ColumnHandle columnHandle)
+    {
+        IcebergColumnHandle column = (IcebergColumnHandle) columnHandle;
+        return new ColumnMetadata(column.getName(), column.getType(), column.getComment().orElse(""), false);
```
nit: hmmm, I'm not sure if we should set a null or use orElse("") here. why do you use an empty string rather than null?

jackye1995 replied:
good point, this is to match the behavior of `IcebergMetadata`, but I agree it seems awkward to return empty string. That constructor is deprecated anyway, I will switch to use the builder.

beinan commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergNativeMetadata.java:
```diff
@@ -0,0 +1,523 @@
+    {
+        return table.schema().columns().stream()
+                .map(column -> {
+                    return new ColumnMetadata(column.name(), toPrestoType(column.type(), typeManager), column.doc(), false);
```
"return" is not necessary I think.  can we just write ".map(column -> new ColumnMetadata(column.name(), toPrestoType(column.type(), typeManager), column.doc(), false))"

jackye1995 replied:
Similar to the last one, this is to match the behavior of `IcebergMetadata`, but that constructor is deprecated anyway, I will switch to use the builder.

beinan reviewed (commented):
Awesome PR!  I heard quite a few giant companies are waiting for this one.  Really appreciate your contribution!   Just put a few comments, I will continue review it tomorrow.

ChunxuTang reviewed (commented):
@jackye1995 Thanks so much for your work! Generally, the PR looks good to me. We can iterate on the work here for further improvement.
I noticed that there're some compilation errors. At your convenience, could you fix them and re-run the tests? Thanks!

jackye1995 commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergConfig.java:
```diff
@@ -52,4 +59,58 @@ public IcebergConfig setCompressionCodec(HiveCompressionCodec compressionCodec)
+        return catalogCacheSize;
+    }
+
+    @Config("iceberg.catalog.cache-size")
```
sure, I will change to `cached-catalog-num` instead, using `-` to match the convention of Presto config keys.

jackye1995 commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergNativeMetadata.java:
```diff
@@ -0,0 +1,523 @@
+            table = resourceFactory.getCatalog(session).loadTable(tableIdentifier);
+        }
+        catch (NoSuchTableException e) {
+            return null;
```
I think returning null if table not exist in `getTableHandle` seems like the consistent behavior across all connectors, so it's probably fine to not give an explanation.

jackye1995 commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergNativeMetadata.java:
```diff
@@ -0,0 +1,523 @@
+        SchemaTableName systemTableName = new SchemaTableName(tableName.getSchemaName(), name.getTableNameWithType());
+        Optional<Long> snapshotId = Optional.of(name.getSnapshotId().orElse(table.currentSnapshot().snapshotId()));
+        switch (name.getTableType()) {
+            case DATA:
```
sure, I moved it together with the default case with a comment.

jackye1995 commented:
@ChunxuTang Thanks for the review! The error was due to rebase, I have fixed it and ran tests locally to make sure it passes. I have also fixed based on your comments.

beinan commented:
rerun the test jobs

beinan reviewed (approved):
lgtm,  great contribution for the presto+iceberg use cases.  thanks a lot!

beinan commented:
> @ChunxuTang Thanks for the review! The error was due to rebase, I have fixed it and ran tests locally to make sure it passes. I have also fixed based on your comments.

@jackye1995 could you rebase master again?  looks like some of the tests are still failing.  Thank you!
